In [1]:
from bs4 import BeautifulSoup
import urllib
from urllib.request import urlopen
import re
import pandas as pd
import numpy as np
import requests
from requests import get
import nltk

# Pulling all links off of search pages 1-30 and putting them in a dataframe to scrape.

In [2]:
# Website for search results for "Data Scientist" in "USA"
base_url=urlopen('https://www.indeed.com/jobs?q=data+scientist&l=USA')
bs = BeautifulSoup(base_url, 'html.parser')

In [6]:
pages = [str(i) for i in range(1,31)]

# Declaring the lists to store data in
Linkdf=[]

# For every page in the interval 1-20 (pages in sets of 9 jobs)
for page in pages:

    page_links = bs.find_all('a', {'href':re.compile("/rc/")})
    for items in page_links:
        Linkdf.append(items['href'])
        print(Linkdf)

['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3']
['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3']
['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3']
['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3']
['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', '/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3']
['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid

['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', '/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3', '/rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3', '/rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3', '/rc/clk?jk=83e3f665b22e5695&fccid=f01127702f08b4dc&vjs=3', '/rc/clk?jk=5726bb7b9a6dfb44&fccid=7077d7e88049c02a&vjs=3', '/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', '/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3', '/rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3', '/rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3', '/rc/clk?jk=83e3f665b22e5695&fccid=f011

In [7]:
len(Linkdf)

270

In [8]:
#Create array with just last one (use the code above to use the right length)
Finallinks=Linkdf[:180]
print(Finallinks)

['/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', '/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3', '/rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3', '/rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3', '/rc/clk?jk=83e3f665b22e5695&fccid=f01127702f08b4dc&vjs=3', '/rc/clk?jk=5726bb7b9a6dfb44&fccid=7077d7e88049c02a&vjs=3', '/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', '/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', '/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', '/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', '/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3', '/rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3', '/rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3', '/rc/clk?jk=83e3f665b22e5695&fccid=f011

In [9]:
#This code allows the code to display the full website instead of truncating
pd.set_option('display.max_colwidth',-1)

#Moving it to a data frame
data = {'links':Finallinks}
df = pd.DataFrame(data, columns=['links'])
df

links
0    /rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3
1    /rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3
2    /rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3
3    /rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3
4    /rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3
..                                                        ...
175  /rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3
176  /rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3
177  /rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3
178  /rc/clk?jk=83e3f665b22e5695&fccid=f01127702f08b4dc&vjs=3
179  /rc/clk?jk=5726bb7b9a6dfb44&fccid=7077d7e88049c02a&vjs=3

[180 rows x 1 columns]

In [10]:
#append indeed.com to the front of each
df['Web'] = 'https://www.indeed.com'
df['URL'] = df.Web.str.cat(df.links)

#sanity check
print(df)

                                                        links  \
0    /rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3   
1    /rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3   
2    /rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3   
3    /rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3   
4    /rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3   
..                                                        ...   
175  /rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3   
176  /rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3   
177  /rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3   
178  /rc/clk?jk=83e3f665b22e5695&fccid=f01127702f08b4dc&vjs=3   
179  /rc/clk?jk=5726bb7b9a6dfb44&fccid=7077d7e88049c02a&vjs=3   

                        Web  \
0    https://www.indeed.com   
1    https://www.indeed.com   
2    https://www.indeed.com   
3    https://www.indeed.com   
4    https://www.indeed.com   
..                      ...   
17

In [11]:
#pull out just a list of the websites.
websites=list(df['URL'])
print(websites)

['https://www.indeed.com/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', 'https://www.indeed.com/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', 'https://www.indeed.com/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', 'https://www.indeed.com/rc/clk?jk=f5dc0ae712778b2a&fccid=87296c430fc584e1&vjs=3', 'https://www.indeed.com/rc/clk?jk=8dcad7369ab9dc70&fccid=3c2181d40a370308&vjs=3', 'https://www.indeed.com/rc/clk?jk=69f9e90c280eb5ee&fccid=4c3bc2bd3fda6e72&vjs=3', 'https://www.indeed.com/rc/clk?jk=0da0d97fcb637195&fccid=29f4096cfef5d289&vjs=3', 'https://www.indeed.com/rc/clk?jk=83e3f665b22e5695&fccid=f01127702f08b4dc&vjs=3', 'https://www.indeed.com/rc/clk?jk=5726bb7b9a6dfb44&fccid=7077d7e88049c02a&vjs=3', 'https://www.indeed.com/rc/clk?jk=267e7cf7de2959c7&fccid=de71a49b535e21cb&vjs=3', 'https://www.indeed.com/rc/clk?jk=75c6f0b9d2079aaf&fccid=9a55d5dc8c10ef41&vjs=3', 'https://www.indeed.com/rc/clk?jk=a11d9b56688c56c9&fccid=847cd08b3d651258&vjs=3', 'https://www.in

# Scraping these 270 websites for Company and Job Description and put into 1 dataframe

In [12]:
#Code from Jodi's NLP homework 5 - adjusting it to just pull reviews now, but can also do tags and chunks if needed

Descriptions=[]
Company=[]

#tags=[]
#chunks=[]
for url in websites:
    response=get(url)
    soup=BeautifulSoup(response.text,'html.parser')
    
    job_containers= soup.find('div', {"class":"jobsearch-jobDescriptionText"})
    title_containers =soup.find('div', {"class":"jobsearch-DesktopStickyContainer"})
    
    job_descriptions=str(job_containers.text)
    job_company=str(title_containers.text)
    
    tokenize_descriptions=nltk.word_tokenize(job_descriptions)
    tokenize_company=nltk.word_tokenize(job_company)
    
    #tagging=nltk.pos_tag(tokenize_reviews)
    
    #chunk=nlp(job_descriptions)
    #spacy_NP=[chunk.text for chunk in chunk.noun_chunks]
    
    Company.append(tokenize_company)
    Descriptions.append(tokenize_descriptions)
    #tags.append(tagging)
    #chunks.append(spacy_NP)

In [13]:
#add all to a dataframe
df_companyinfo = pd.DataFrame({'Company_Info': Company})
df_companyinfo

Company_Info
0    [Entry, Level, Data, ScientistIBM26,459, reviews-United, StatesIBM26,459, reviewsRead, what, people, are, saying, about, working, here, .]                                                        
1    [Junior, Data, ScientistLowe, 's, Inc.34,252, reviews-Mooresville, ,, NCLowe, 's, Inc.34,252, reviewsRead, what, people, are, saying, about, working, here, .]                                    
2    [Data, ScientistValen, AnalyticsDenver, ,, CO]                                                                                                                                                    
3    [Data, ScientistSony, Pictures, Entertainment, Inc.360, reviews-Culver, City, ,, CA, 90232Sony, Pictures, Entertainment, Inc.360, reviewsRead, what, people, are, saying, about, working, here, .]
4    [Junior, Data, ScientistBlueVine-Redwood, City, ,, CA]                                                                                                                                            
..                                                      ...                                                                                                                                            
175  [Junior, Data, ScientistBlueVine-Redwood, City, ,, CA]                                                                                                                                            
176  [Data, ScientistSocial, Science, Research, Solutions, (, SSRS, ), -Glen, Mills, ,, PA]                                                                                                            
177  [Data, ScientistNew, York, City, DEPT, OF, INFO, TECH, &, TELECOMMManhattan, ,, NY, $, 63,031, -, $, 145,000, a, year]                                                                            
178  [Data, ScientistDigitalogy-New, York, ,, NY, $, 50, -, $, 80, an, hour, -, Contract]                                                                                                              
179  [Data, ScientistAETNA4,486, reviews-New, York, ,, NY, 10016AETNA4,486, reviewsRead, what, people, are, saying, about, working, here, .]                                                           

[180 rows x 1 columns]

In [14]:
#All ad dataframe
df_descriptions = pd.DataFrame({'Full_Description': Descriptions})
df_descriptions

Full_Description
0    [Introduction, At, IBM, we, have, an, amazing, opportunity, to, transform, the, world, with, cognitive, technology, ., By, using, the, vast, amounts, of, information, available, today, to, identify, new, patterns, and, make, new, discoveries, ,, we, are, helping, cities, become, smarter, ,, hospitals, transform, patient, care, ,, financial, institutions, minimize, risk, ,, and, pharmaceuticals, find, cures, for, rare, diseases, ., Data, scientists, work, with, enterprise, leaders, and, key, decision, makers, to, solve, business, problems, by, preparing, ,, analyzing, ,, and, understanding, data, to, deliver, insight, ,, predict, emerging, trends, ,, and, provide, recommendations, to, optimize, results, ., Data, scientists, use, a, ...]              
1    [Mines, and, extracts, data, and, applies, statistics, and, algorithms, necessary, to, derive, insights, for, Lowe, 's, and, Lowe, 's, Digital, ., Supports, the, creation, of, an, automated, insights, generation, framework, for, Lowe, 's, business, partners, to, effectively, interpret, data, ., Provides, actionable, insights, through, data, science, on, Personalization, ,, Search, &, Navigation, ,, SEO, &, Promotions, ,, Supply, Chain, ,, Services, ,, other, company, priorities, ,, etc, ., Tracks, success, of, Lowe, 's, through, the, development, of, dashboard, reports, that, measure, financial, results, ,, customer, satisfaction, ,, and, engagement, metrics, ., Conducts, deep, statistical, analysis, ,, including, predictive, and, prescriptive, ...]
2    [Join, a, high, performing, and, rapidly, growing, team, Valen, Analytics, is, a, rapidly, expanding, advanced, data, and, predictive, analytics, company, headquartered, in, downtown, Denver, ., Valen, ’, s, state-of-the-art, analytics, and, predictive, modeling, products, are, built, on, Valen, ’, s, unique, industry-wide, consortium, data, platform, ,, specifically, designed, for, property, and, casualty, insurance, carriers, ., If, you, want, to, help, build, an, industry-changing, company, and, you, thrive, in, a, fast-paced, environment, where, the, only, rule, is, to, WOW, customers, ,, colleagues, and, shareholders, ,, then, we, encourage, you, to, submit, your, resume, ., This, is, not, a, job, ., It, ...]                                    
3    [Data, Scientist, Sony, Pictures, Entertainment, is, an, equal, opportunity, employer, ., We, evaluate, qualified, applicants, without, regard, to, race, ,, color, ,, religion, ,, sex, ,, national, origin, ,, disability, ,, veteran, status, ,, age, ,, sexual, orientation, ,, gender, identity, ,, or, other, protected, characteristics, ., USA, -, Culver, City, -, Studio, Lot, -, Norman, Lear, Bldg]                                                                                                                                                                                                                                                                                                                                                                        
4    [Small, business, owners, don, ’, t, get, the, love, they, deserve, from, traditional, banks, ., That, ’, s, where, BlueVine, comes, in, ., BlueVine, is, a, leading, provider, of, online, working, capital, financing, to, small, and, mid-sized, businesses, ., We, ’, re, using, cutting, edge, technology, to, disrupt, the, traditional, banking, industry, and, develop, the, next, generation, of, fast, and, simple, financial, products, ,, designed, for, today, ’, s, small, businesses, ., BlueVine, is, headquartered, in, Redwood, City, ,, CA, and, backed, by, leading, investors, including, Menlo, Ventures, ,, Lightspeed, Ventures, ,, Citi, Ventures, ,, SVB, Financial, ,, Nationwide, Insurance, ...]                                                          
..                                                                                                                                                            

In [15]:
FullDataSet = pd.concat([df_companyinfo, df_descriptions], axis=1)
FullDataSet

Company_Info  \
0    [Entry, Level, Data, ScientistIBM26,459, reviews-United, StatesIBM26,459, reviewsRead, what, people, are, saying, about, working, here, .]                                                           
1    [Junior, Data, ScientistLowe, 's, Inc.34,252, reviews-Mooresville, ,, NCLowe, 's, Inc.34,252, reviewsRead, what, people, are, saying, about, working, here, .]                                       
2    [Data, ScientistValen, AnalyticsDenver, ,, CO]                                                                                                                                                       
3    [Data, ScientistSony, Pictures, Entertainment, Inc.360, reviews-Culver, City, ,, CA, 90232Sony, Pictures, Entertainment, Inc.360, reviewsRead, what, people, are, saying, about, working, here, .]   
4    [Junior, Data, ScientistBlueVine-Redwood, City, ,, CA]                                                                                                                                               
..                                                      ...                                                                                                                                               
175  [Junior, Data, ScientistBlueVine-Redwood, City, ,, CA]                                                                                                                                               
176  [Data, ScientistSocial, Science, Research, Solutions, (, SSRS, ), -Glen, Mills, ,, PA]                                                                                                               
177  [Data, ScientistNew, York, City, DEPT, OF, INFO, TECH, &, TELECOMMManhattan, ,, NY, $, 63,031, -, $, 145,000, a, year]                                                                               
178  [Data, ScientistDigitalogy-New, York, ,, NY, $, 50, -, $, 80, an, hour, -, Contract]                                                                                                                 
179  [Data, ScientistAETNA4,486, reviews-New, York, ,, NY, 10016AETNA4,486, reviewsRead, what, people, are, saying, about, working, here, .]                                                              

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Full_Description  
0    [Introduction, At, IBM, we, have, an, amazing, opportunity, to, transform, the, world, with, cognitive, technology, ., By, using, the, vast, amounts, of, information, available, today, to, identify, new, patterns, and, make, new, discoveries, ,, we, are, helping, cities, become, smarter, ,, hospitals, transform, patient, care, ,, financial, institutions, minimize, risk, ,, and, pharmaceuticals, find, cures, for, rare, diseases, ., Data, scientists, work, with, enterprise, leaders, and, key, decision, makers, to, solve, business, problems, by, preparing, ,, analyzing, ,, and, understanding, data, to, deliver, insight, ,, predict, emerging, trends, ,, and, provide, recommendations, to, optimize, results, ., Data, scientists, use, a, ...]                
1    [Mines, and, extracts, data, and, applies, statistics, and, algorithms, necessary, to, derive, insights, for, Lowe, 's, and, Lowe, 's, Digital, ., Supports, the, creation, of, an, automated, insights, generation